pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/ReplicaGroupInstanceSelector.java

```diff
@@ -156,4 +159,20 @@ private void selectServersUsingAdaptiverServerSelector(List<String> segments, in
       segmentToSelectedInstanceMap.put(segment, selectedInstance);
     }
   }
+
+  private List<String> fetchCandidateServersForQuery(List<String> segments,
+      Map<String, List<String>> segmentToEnabledInstancesMap) {
+    List<String> serversList = new ArrayList<>();
+
+    Set<String> tempServerSet = new HashSet<>();
+    for (String segment : segments) {
+      List<String> enabledInstances = segmentToEnabledInstancesMap.get(segment);
+      if (enabledInstances != null) {
+        tempServerSet.addAll(enabledInstances);
+      }
+    }
+
+    serversList.addAll(tempServerSet);
+    return serversList;
+  }
```

Comment:
hmm I am confused why should this code need to be implemented. 

The logic used by replica group instance selector to pick candidate instances first is not getting changes by adaptive selection as the latter runs on top of it so shouldn't this logic already exist in the replica group instance selector ?

Reply:
It already exists. But they way the current code is written is, we do only one loop over segments. For each segment, we pick the server from the list of servers for the segment. 

With AdaptiveServerSelection, we do an initial loop before to determine the list of servers for all segments that are touched by a query.  Then a ranking on them is performed. And then loop over the segments again and now determine which server to pick for a segment.

Reply:
Got it. Thanks for clarifying